File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        //base case
        if(root==NULL){
            return 0;
        }

        int maxsum = INT_MIN;
        maxpathsumhelp(root, maxsum);
        return maxsum;
    }
private:
    int maxpathsumhelp(TreeNode* node, int &maxsum){
        if(node==NULL){
            return 0;
        }

        int leftsum = max(0, maxpathsumhelp(node->left, maxsum));
        int rightsum = max(0, maxpathsumhelp(node->right, maxsum));

        int sumvalue = node->val + leftsum + rightsum;

        maxsum = max(maxsum, sumvalue);

        return node->val + max(leftsum, rightsum);
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def maxpathsumhelper(node: TreeNode) -> int:
            nonlocal maxsum
            if not node:
                return 0

            leftsum = max(0, maxpathsumhelper(node.left))
            rightsum = max(0, maxpathsumhelper(node.right))

            sumvalue = node.val + leftsum + rightsum

            maxsum = max(maxsum, sumvalue)

            return node.val + max(leftsum, rightsum)
        
        maxsum = float('-inf')
        maxpathsumhelper(root)
        return maxsum